Add smile support to coordinator-to-worker protocol

nezihyigitbasi opened the pull request:
Depends on https://github.com/prestodb/presto/pull/12363, so only 1614354 should be reviewed.

cemcayiroglu commented on presto-main/src/main/java/com/facebook/presto/server/HttpRemoteTaskFactory.java:
```diff
@@ -71,29 +74,43 @@
+            JsonCodec<TaskInfo> taskInfoJsonCodec,
+            SmileCodec<TaskInfo> taskInfoSmileCodec,
+            JsonCodec<TaskUpdateRequest> taskUpdateRequestJsonCodec,
+            SmileCodec<TaskUpdateRequest> taskUpdateRequestSmileCodec,
```
Same here. This is confusing. The caller should pick one Codec and pass.

cemcayiroglu commented on presto-main/src/main/java/com/facebook/presto/server/remotetask/ContinuousTaskStatusFetcher.java:
```diff
@@ -142,15 +152,22 @@ private synchronized void scheduleNextRequest()
+        else {
+            responseHandler = createAdaptingJsonResponseHandler(unwrapJsonCodec(taskStatusCodec));
+        }
+
```
Can we pass a response handler based on the config? Instead of doing a if check.

nezihyigitbasi replied:
In theory we can, but not sure what the additional benefit will be.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/memory/RemoteNodeMemory.java:
```diff
@@ -137,4 +157,24 @@ public void onFailure(Throwable t)
+        return jsonBodyGenerator(unwrapJsonCodec(assignmentsRequestCodec), assignments);
+    }
+
+    private Builder setBinaryTransportHeadersIfNeeded(Builder requestBuilder)
```
We should think of how can this be a part of the `BodyGenerator`. Maybe we should introduce a slightly different interface, that can both, encode the response and set the headers.

arhimondr replied:
Extract it into a helper method

arhimondr commented:
@nezihyigitbasi As i understand it is only a temporary code, before the Airlift PR is merged.

nezihyigitbasi commented:
> Personally i would try to abstract the `Codec` interface

The Airlift PR has the `Codec` interface: https://github.com/airlift/airlift/pull/706/files#diff-993e22557305cb1d7bf61220f27538c3

> As i understand it is only a temporary code, before the Airlift PR is merged.

Yes, once the Airlift PR is in this change will be cleaned up.

nezihyigitbasi commented on presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryManager.java:
```diff
@@ -132,32 +137,43 @@ public ClusterMemoryManager(
             MBeanExporter exporter,
-            JsonCodec<MemoryInfo> memoryInfoCodec,
+            JsonCodec<MemoryInfo> memoryInfoJsonCodec,
+            SmileCodec<MemoryInfo> memoryInfoSmileCodec,
```
Yep, these are injected by Guice. The Airlift PR has a `Codec` interface and that would simplify this code.

arhimondr commented:
> The Airlift PR has the `Codec` interface

Yeah, but we should add the other method there as well. Currently it has only `byte[] toBytes(T instance)`, but there should be also `T fromBytes(byte[] data)`. 

When having both of these method a lot of interfaces can be simplified, and a lot of copied classed in the Airlift PR can be simplified.

Could you please address these comments in the Airlift PR? I think once the comments are addressed that PR will be a good fit for the Arilift.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/server/remotetask/SimpleHttpResponseHandler.java:
```diff
@@ -79,6 +77,23 @@ else if (response.getStatusCode() == HttpStatus.SERVICE_UNAVAILABLE.code()) {
+                    response.getStatusMessage(),
+                    unwrapJsonResponse(response).getResponseBody());
+        }
+        return format("Expected response code from %s to be %s, but was %s: %s",
```
Should we print the binary json (say as base64)?

nezihyigitbasi replied:
I am not sure how useful that would be for debugging.

arhimondr commented on presto-tests/src/test/java/com/facebook/presto/tests/TestSmileSupport.java:
```diff
@@ -0,0 +1,31 @@
+        try (DistributedQueryRunner queryRunner = TpchQueryRunnerBuilder.builder()
+                .setSingleExtraProperty("experimental.internal-communication.binary-transport-enabled", "true")
+                .build()) {
+            queryRunner.execute("SELECT * FROM lineitem LIMIT 10");
```
Maybe try to run something bigger here, that has more stages and tasks

nezihyigitbasi commented:
Addressed your comments, thanks for the review. I also addressed the comments in Airlift PR. Once that's merged these changes will get cleaned up & depend on the implementation in Airlift.

tdcmeehan commented:
> Yes, once the Airlift PR is in this change will be cleaned up.

While I trust that this will be taken care of, I think there should be a task for tracking (I couldn't find one).

nezihyigitbasi commented:
@tdcmeehan Created an issue to track it: https://github.com/prestodb/presto/issues/12411
